### PR TITLE
[SPARK-8727][SQL] Missing python api; md5, log2, pi, e

### PR DESCRIPTION
**Pull request by tarekbecker**
Jira: https://issues.apache.org/jira/browse/SPARK-8727

### CODE REVIEW

**Comment by JoshRosen** [comment]
Jenkins, this is ok to test.

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -341,6 +371,20 @@ def monotonicallyIncreasingId():
     return Column(sc._jvm.functions.monotonicallyIncreasingId())
 
 
+@since(1.5)
+def pi():
+    """
+    Returns the float value that is closer than any other to pi, the ratio of the circumference
+    of a circle to its diameter.
+
+    >>> sqlContext.createDataFrame([(1,)], ['a']).select((pi()).alias('pi')).collect()
```
Comment (tarekauel) [inline python/pyspark/sql/functions.py]: Is it possible to do this more concise? `sqlContext.createDataFrame([(pi(),)], ['a']).collect()` fails: `cannot create an RDD from type: <type 'list'>`. Sorry if the answer is obvious, I am not a python geek.

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -383,19 +440,6 @@ def sha2(col, numBits):
     return Column(jc)
 
 
-@ignore_unicode_prefix
-@since(1.5)
-def sha1(col):
```
Comment (tarekauel) [inline python/pyspark/sql/functions.py]: FYI: I haven't changed the method, I just fixed the order: sha1, sha2

**Comment by tarekauel** [comment]
@JoshRosen 
On my local machine runs `Python 2.7.9 :: Anaconda 2.2.0 (x86_64)` and I get `[Row(e=2.718281828459045)]` as result. But Jenkins seems to get `[Row(e=2.7182818284590451)]`, looks like a rounding error. Shall we round the result, in order to avoid problems in the future?

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -299,6 +302,22 @@ def countDistinct(col, *cols):
     return Column(jc)
 
 
+@since(1.5)
+def e():
+    """
+    Returns the float value that is closer than any other to e, the base of the natural
+    logarithms.
+
+    >>> e = sqlContext.createDataFrame([(1,)], ['a']).select((e()).alias('e')).collect()
```
Comment (davies) [inline python/pyspark/sql/functions.py]: You can use the default `df`, also first() will be better than `collect()` in this case.

File: python/pyspark/sql/functions.py [hunk 4]
```diff
@@ -471,6 +531,17 @@ def log(arg1, arg2=None):
     return Column(jc)
 
 
+@since(1.5)
+def log2(col):
+    """Returns the base-2 logarithm of the argument.
+
+    >>> sqlContext.createDataFrame([(4,)], ['a']).select(log2('a').alias('log2')).collect()
```
Comment (davies) [inline python/pyspark/sql/functions.py]: `['a']` is not necessary

File: python/pyspark/sql/functions.py [hunk 5]
```diff
@@ -410,6 +458,18 @@ def sparkPartitionId():
...
 @ignore_unicode_prefix
+@since(1.5)
+def strlen(col):
+    """Calculates the length of a string expression.
+
+    >>> sqlContext.createDataFrame([('ABC',)], ['a']).select(strlen('a').alias('e')).collect()
+    [Row(e=3)]
+    """
+    sc = SparkContext._active_spark_context
+    return Column(sc._jvm.functions.strlen(col))
```
Comment (davies) [inline python/pyspark/sql/functions.py]: Should we check `col` is Column object or not?

File: python/pyspark/sql/functions.py (hunk 4 above)
Comment (davies) [inline python/pyspark/sql/functions.py]: nvm, it's needed.

File: python/pyspark/sql/functions.py [hunk 6]
```diff
@@ -299,6 +302,22 @@ def countDistinct(col, *cols):
     return Column(jc)
 
 
+@since(1.5)
+def e():
```
Comment (davies) [inline python/pyspark/sql/functions.py]: In python, we have `math.e` and `math.pi`, is this still useful? cc @rxin

**Comment by davies** [comment]
@tarekauel Could you update the title to include `[SQL]` and function names?

File: python/pyspark/sql/functions.py (hunk 6 above)
Comment (rxin) [inline python/pyspark/sql/functions.py]: actually even in java/scala it is probably not as useful. i think we should just remove those from the dataframe api, but leave the expression there for sql.

Comment (davies) [inline python/pyspark/sql/functions.py]: +1

Comment (tarekauel) [inline python/pyspark/sql/functions.py]: Shall I remove it from the dataframe api in this pr?

Comment (davies) [inline python/pyspark/sql/functions.py]: Good to have.

**Comment by rxin** [comment]
I created https://github.com/apache/spark/pull/7137 to remove e/pi from Scala/Java.

**Comment by davies** [comment]
Merging this into master (updated the title), thanks!

**Pull request closed without merging**